src/internal_modules/roc_fec/reader.cpp

```diff
@@ -59,26 +60,35 @@ bool Reader::is_alive() const {
     return alive_;
 }
 
-packet::PacketPtr Reader::read() {
+int Reader::read(packet::PacketPtr& pp) {
     roc_panic_if_not(is_valid());
+
     if (!alive_) {
-        return NULL;
+        return error::ErrNoData;
     }
-    packet::PacketPtr pp = read_();
-    if (pp) {
+
+    int code = read_(pp);
+    if (code == 0) {
         n_packets_++;
     }
-    // check if alive_ has changed
-    return (alive_ ? pp : NULL);
+    if (!alive_) {
+        code = error::ErrNoData;
+        pp = NULL;
+    }
+
+    return code;
 }
 
-packet::PacketPtr Reader::read_() {
-    fetch_packets_();
+int Reader::read_(packet::PacketPtr& ptr) {
+    const int code = fetch_all_packets_();
+    if (code < 0) {
+        return code;
+    }
 
     if (!started_) {
         packet::PacketPtr pp = get_first_packet_();
         if (!pp || pp->fec()->encoding_symbol_id > 0) {
-            return source_queue_.read();
+            return source_queue_.read(ptr);
         }
 
         roc_log(LogDebug,
```

Comment:
I suggest small refactoring for this function:

```cpp
int Reader::read_(packet::PacketPtr& ptr) {
    const int code = fetch_all_packets_();
    if (code < 0) {
        return code;
    }

    if (!started_) {
        started_ = try_start();
    }

    if (!started_) {
        // until started, just forward all source packets
        return source_queue_.read(ptr);
    }

    return get_next_packet_(ptr);
}
```

Here, `get_first_packet_()` is renamed to `try_start_()` and now it returns bool.

This check:

```
if (!pp || pp->fec()->encoding_symbol_id > 0) {
```

is moved inside `try_start_()`.

Also please add a comment for `pp->fec()->encoding_symbol_id > 0` (it will be now inside `try_start_()`):

```cpp
// wait until we receive first packet in block (ESI=0)
// see also gh-186
if (pp->fec()->encoding_symbol_id > 0)
```

This piece:

```cpp
        roc_log(LogDebug,
                "fec reader: got first packet in a block, start decoding:"
                " n_packets_before=%u sbn=%lu",
                n_packets_, (unsigned long)cur_sbn_);
```

is also moved into `try_start_()`.